README.md
![Visitor Badge](https://visitor-badge.laobi.icu/badge?page_id=rust-ufw-app-profile)
![Crates Badge](https://img.shields.io/crates/v/ufwprofile)
![Crates Downloads](https://img.shields.io/crates/d/ufwprofile)

# About Project

UFW app profile written in pure rust.

## Implementation:

```toml
ufwprofile = "" #check latest version above
```

Or

```
cargo add ufwprofile
```

# Example

```rust
fn main() -> anyhow::Result<()> {
    if ufwprofile::UFWConf::check_write_permission() {
        //checks if ufw exists and the path /etc/ufw/applications.d is writable
        let conf = ufwprofile::UFWConf::init("AppName", "Title", "Description")?
            .append_ports("80", "")?
            .append_ports("81:82", "tcp")?
            .append_ports("84", "udp")?
            .append_ports("83", "")?
            .append_ports("8000", "tcp")?;

        if ufwprofile::UFWConf::is_root() {
            // check if the app has root permission.
            println!("{}", conf.try_adding_to_ufw(true).unwrap());
        } else {
            println!("{}", conf.try_write_with_sudo(true).unwrap());
        }
    } else {
        println!("Unable to write");
    }
    Ok(())
}
```

# Drawbacks

### This is hardcoded dependency

1. The config file is hardcoded.
2. The path is assumed to be `/etc/ufw/applications.d`

# Changelog

check [CHANGELOG.md](CHANGELOG.md)


CHANGELOG.md
# 0.1.0

- initial release

# 0.2.\*

- added check_permissions(), try_write() and get_config_string()

# 1.3.\*

- changed `check_permissions()` to `check_write_permission()` and added `is_root()` method to check if the app has admin privileges


examples/main.rs
#[cfg(test)]
mod tests {
    use std::fs::File;
    use std::io::Write;
    use std::process::Command;

    #[test]
    fn foo() {
        let x = match File::create("foo.txt") {
            Ok(mut f) => match f.write_all(b"alo") {
                Ok(_) => true,
                Err(_) => false,
            },
            Err(_) => false,
        };
        println!("{x}");
    }

    #[test]
    fn bar() {
        let x = match File::create("/etc/ufw/applications.d/foo".to_string().as_str()) {
            Ok(mut f) => match f.write_all(b"alo") {
                Ok(_) => {
                    let mut x = Command::new("ufw");
                    match x.arg("allow").arg("foo").output() {
                        Ok(_) => true,
                        Err(_) => false,
                    }
                }
                Err(_) => false,
            },
            Err(_) => false,
        };
    }
}

fn main() -> anyhow::Result<()> {
    if ufwprofile::UFWConf::check_write_permission() {
        //checks if ufw exists and the path /etc/ufw/applications.d is writable
        let conf = ufwprofile::UFWConf::init("AppName", "Title", "Description")?
            .append_ports("80", "")?
            .append_ports("81:82", "tcp")?
            .append_ports("84", "udp")?
            .append_ports("83", "")?
            .append_ports("8000", "tcp")?;

        if ufwprofile::UFWConf::is_root() {
            // check if the app has root permission.
            println!("{}", conf.try_adding_to_ufw(true).unwrap());
        } else {
            println!("{}", conf.try_write_with_sudo(true).unwrap());
        }
    } else {
        println!("Unable to write");
    }
    Ok(())
}


src/rootcheck.rs
//! [![crates.io](https://img.shields.io/crates/v/sudo?logo=rust)](https://crates.io/crates/sudo/)
//! [![docs.rs](https://docs.rs/sudo/badge.svg)](https://docs.rs/sudo)
//!
//! Detect if you are running as root, restart self with `sudo` if needed or setup uid zero when running with the SUID flag set.
//!
//! ## Requirements
//!
//! * The `sudo` program is required to be installed and setup correctly on the target system.
//! * Linux or Mac OS X tested
//!     * It should work on *BSD. However, it is not tested.
#![allow(clippy::bool_comparison)]

use libc::setuid;
use log::trace;
use std::process::Command;

/// Cross platform representation of the state the current program running
#[derive(Debug, PartialEq)]
pub enum RunningAs {
    /// Root (Linux/Mac OS/Unix) or Administrator (Windows)
    Root,
    /// Running as a normal user
    User,
    /// Started from SUID, a call to `sudo::escalate_if_needed` or `sudo::with_env` is required to claim the root privileges at runtime.
    /// This does not restart the process.
    Suid,
}
use RunningAs::*;

#[cfg(unix)]
/// Check getuid() and geteuid() to learn about the configuration this program is running under
pub fn check() -> RunningAs {
    let uid = unsafe { libc::getuid() };
    let euid = unsafe { libc::geteuid() };

    match (uid, euid) {
        (0, 0) => Root,
        (_, 0) => Suid,
        (_, _) => User,
    }
    //if uid == 0 { Root } else { User }
}

#[cfg(unix)]
#[inline]
pub fn escalate_if_needed() -> bool {
    with_env(&[])
}

#[cfg(unix)]
pub fn with_env(prefixes: &[&str]) -> bool {
    let current = check();
    trace!("Running as {:?}", current);
    match current {
        Root => {
            trace!("already running as Root");
            return true;
            // return Ok(current);
        }
        Suid => {
            trace!("setuid(0)");
            unsafe {
                setuid(0);
            }
            return true;
            // return Ok(current);
        }
        User => {
            log::debug!("Escalating privileges");
        }
    }

    let mut args: Vec<_> = std::env::args().collect();
    if let Some(absolute_path) = std::env::current_exe()
        .ok()
        .and_then(|p| p.to_str().map(|p| p.to_string()))
    {
        args[0] = absolute_path;
    }
    let mut command: Command = Command::new("/usr/bin/sudo");

    // Always propagate RUST_BACKTRACE
    if let Ok(trace) = std::env::var("RUST_BACKTRACE") {
        let value = match &*trace.to_lowercase() {
            "" => None,
            "1" | "true" => Some("1"),
            "full" => Some("full"),
            invalid => {
                log::warn!(
                    "RUST_BACKTRACE has invalid value {:?} -> defaulting to \"full\"",
                    invalid
                );
                Some("full")
            }
        };
        if let Some(value) = value {
            trace!("relaying RUST_BACKTRACE={}", value);
            command.arg(format!("RUST_BACKTRACE={}", value));
        }
    }

    if prefixes.is_empty() == false {
        for (name, value) in std::env::vars().filter(|(name, _)| name != "RUST_BACKTRACE") {
            if prefixes.iter().any(|prefix| name.starts_with(prefix)) {
                trace!("propagating {}={}", name, value);
                command.arg(format!("{}={}", name, value));
            }
        }
    }

    let mut child = command.args(args).spawn().expect("failed to execute child");

    let ecode = child.wait().expect("failed to wait on child");

    ecode.success()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let _ = check();
    }
}


src/config.rs
use crate::rootcheck;
use anyhow::{anyhow, Result};
use regex::Regex;
use std::fs::File;
use std::io::{Error, ErrorKind, Write};
use std::path::Path;
use std::process::Command;

///  Struct that contains app name, config string, ports HashMap
/// Example:
/// ```
/// fn main() -> anyhow::Result<()> {
///     if ufwprofile::UFWConf::check_write_permission() {
///         //checks if ufw exists and the path /etc/ufw/applications.d is writable
///         let conf = ufwprofile::UFWConf::init("AppName", "Title", "Description")?
///             .append_ports("80", "")?
///             .append_ports("81:82", "tcp")?
///             .append_ports("84", "udp")?
///             .append_ports("83", "")?
///             .append_ports("8000", "tcp")?;
///
///         if ufwprofile::UFWConf::is_root() {
///             // check if the app has root permission.
///             println!("{}", conf.try_adding_to_ufw(true).unwrap());
///         } else {
///             println!("{}", conf.try_write_with_sudo(true).unwrap());
///         }
///     }
///     Ok(())
/// }
/// ```
#[derive(Default, Clone)]
pub struct UFWConf {
    app_name: String,
    config: String,
}

impl UFWConf {
    pub fn is_root() -> bool {
        rootcheck::escalate_if_needed()
    }

    /// To add ports to the config, pass `port` and `protocol` to be added
    ///
    /// You can pass empty string to allow all protocols
    pub fn append_ports(&mut self, port: &str, protocol: &str) -> Result<Self> {
        let cfg = format_ports(port, protocol)?;
        if self.config.chars().last().unwrap().eq(&'|') {
            self.config.pop();
        }
        self.config = format!("{}|{}", self.config, cfg);
        Ok(self.clone())
    }

    /// check required permissions and if ufw is installed
    pub fn check_write_permission() -> bool {
        match Command::new("ufw").arg("version").spawn() {
            Ok(_) => !std::fs::metadata("/etc/ufw/applications.d/")
                .unwrap()
                .permissions()
                .readonly(),
            Err(_) => false,
        }
    }

    /// Pass `app_name`, `title` and `description` of the app.
    pub fn init(app_name: &str, title: &str, description: &str) -> Result<Self> {
        let app_name = app_name.replace(' ', "");
        let config = format!(
            "[{}]\ntitle={}\ndescription={}\nports=\n",
            app_name, title, description,
        );
        Ok(Self { app_name, config })
    }

    /// returns config string.
    pub fn get_config_string(&self) -> String {
        self.config.clone()
    }
    /// tries writing it to `/etc/ufw/applications.d/`
    pub fn try_write(&self) -> Result<(), Error> {
        let path = format!("/etc/ufw/applications.d/ufw-{}", self.app_name);
        if Path::new(path.as_str()).exists() {
            std::fs::remove_file(path.as_str()).unwrap();
        }

        match File::create(path) {
            Ok(mut f) => match f.write_all(self.config.as_bytes()) {
                Ok(_) => Ok(()),
                Err(e) => Err(e),
            },
            Err(e) => Err(e),
        }
    }

    /// pass `true` if you want to ALLOW the ports and `false` to DENY the ports.
    pub fn try_write_with_sudo(&self, allow: bool) -> Result<String, Error> {
        let path = format!("/etc/ufw/applications.d/ufw-{}", self.app_name);
        if Path::new(path.as_str()).exists() {
            std::fs::remove_file(path.as_str()).unwrap();
        }
        match File::create(path) {
            Ok(mut f) => match f.write_all(self.config.as_bytes()) {
                Ok(_) => {
                    let mut x = Command::new("sudo");
                    match allow {
                        true => {
                            match x
                                .arg("ufw")
                                .arg("allow")
                                .arg(self.app_name.clone())
                                .output()
                            {
                                Ok(d) => Ok(String::from_utf8(d.stdout).unwrap()),
                                Err(e) => Err(e),
                            }
                        }
                        false => {
                            match x.arg("ufw").arg("deny").arg(self.app_name.clone()).output() {
                                Ok(d) => Ok(String::from_utf8(d.stdout).unwrap()),
                                Err(e) => Err(e),
                            }
                        }
                    }
                }
                Err(e) => Err(Error::new(
                    ErrorKind::Other,
                    format!("Error writing file {}", e),
                )),
            },
            Err(e) => Err(Error::new(
                ErrorKind::Other,
                format!("Error creating file {}", e),
            )),
        }
    }

    /// pass `true` if you want to ALLOW the ports and `false` to DENY the ports.
    pub fn try_adding_to_ufw(&self, allow: bool) -> Result<String> {
        let path = format!("/etc/ufw/applications.d/ufw-{}", self.app_name);
        if Path::new(path.as_str()).exists() {
            std::fs::remove_file(path.as_str()).unwrap();
        }
        match File::create(path) {
            Ok(mut f) => match f.write_all(self.config.as_bytes()) {
                Ok(_) => {
                    let mut x = Command::new("ufw");
                    match allow {
                        true => match x.arg("allow").arg(self.app_name.clone()).output() {
                            Ok(d) => Ok(String::from_utf8(d.stdout).unwrap()),
                            Err(_) => {
                                let mut x = Command::new("sudo");
                                match x
                                    .arg("ufw")
                                    .arg("allow")
                                    .arg(self.app_name.clone())
                                    .output()
                                {
                                    Ok(d) => Ok(String::from_utf8(d.stdout).unwrap()),
                                    Err(e) => Err(anyhow!(format!(
                                        "Error running ufw deny {}: {}",
                                        self.app_name, e
                                    ),)),
                                }
                            }
                        },
                        false => match x.arg("deny").arg(self.app_name.clone()).output() {
                            Ok(d) => Ok(String::from_utf8(d.stdout).unwrap()),
                            Err(e) => Err(anyhow!("{}", e)),
                        },
                    }
                }
                Err(e) => Err(anyhow!(format!("Error writing file {}", e),)),
            },
            Err(e) => Err(anyhow!(format!("Error creating file {}", e),)),
        }
    }
}

fn format_ports(port: &str, protocol: &str) -> Result<String> {
    check_ports(port, protocol)?;
    let protocol = if protocol.is_empty() {
        String::new()
    } else {
        format!("/{protocol}")
    };
    Ok(format!("{port}{protocol}|"))
}

#[inline]
fn check_ports(port_range: &str, protocol: &str) -> Result<()> {
    let port_pattern = Regex::new(r"^\d+:\d+$")?;
    if !["tcp", "udp", ""].contains(&protocol) {
        return Err(anyhow!("Invalid protocol {port_range}={protocol}"));
    }
    if !port_pattern.is_match(port_range) {
        return Err(anyhow!("Bad port/range at {port_range}={protocol}"));
    }
    Ok(())
}
